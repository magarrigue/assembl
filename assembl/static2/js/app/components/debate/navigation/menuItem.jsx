// @flow
import * as React from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { isMobile } from '../../../utils/globalFunctions';
import { get, goTo } from '../../../utils/routeMap';
import PostsAndContributorsCount from '../../common/postsAndContributorsCount';

type ItemNode = {
  id: string,
  title: string,
  img: {
    externalUrl: string
  },
  numContributors: number,
  numPosts: number
};

type MenuItemProps = {
  item: ItemNode,
  identifier: string,
  phaseId: string,
  selected: boolean,
  hasSubItems: boolean,
  slug: string,
  onClick: Function,
  onMouseOver: Function,
  onMouseLeave: Function
};

export class DumbMenuItem extends React.Component<MenuItemProps> {
  showMenu = () => {
    const { onMouseOver, item } = this.props;
    if (onMouseOver) onMouseOver(item.id);
  };

  hideMenu = () => {
    const { onMouseLeave, item } = this.props;
    if (onMouseLeave) onMouseLeave(item.id);
  };

  onLinkClick = () => {
    const { identifier, phaseId, item, slug, onClick } = this.props;
    if (onClick) onClick();
    goTo(get('idea', { slug: slug, phase: identifier, phaseId: phaseId, themeId: item.id }));
  };

  render() {
    const { item, selected, hasSubItems } = this.props;
    const { title, img, numContributors, numPosts } = item;
    // The first touch show the menu and the second activate the link
    const isTouchScreenDevice = isMobile.any();
    const displayArrow = selected && hasSubItems;
    const touchActive = isTouchScreenDevice && hasSubItems && !selected;
    const onLinkClick = touchActive ? this.showMenu : this.onLinkClick;
    return (
      <div
        className={classNames('menu-item-container', {
          active: selected,
          empty: selected && !displayArrow
        })}
        onMouseOver={!isTouchScreenDevice && this.showMenu}
        onMouseLeave={!isTouchScreenDevice && this.hideMenu}
      >
        <div onClick={onLinkClick} className="menu-item">
          <div className="thumb-img" style={img && img.externalUrl ? { backgroundImage: `url(${img.externalUrl})` } : null}>
            <div className="thumb-img-background" />
          </div>
          <div className="thumb-body">
            <div title={title} className="thumb-title">
              {title}
            </div>
            <PostsAndContributorsCount className="menu-stats" numContributors={numContributors} numPosts={numPosts} />
          </div>
          {displayArrow && <span className="thumb-arrow assembl-icon assembl-icon-right-dir" />}
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  slug: state.debate.debateData.slug
});

export default connect(mapStateToProps)(DumbMenuItem);