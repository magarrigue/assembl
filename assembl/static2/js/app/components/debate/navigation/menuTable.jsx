// @flow
import * as React from 'react';
import type { ApolloClient } from 'react-apollo';

import { SurveyTable, IdeasTable } from './tables';
import AllIdeasQuery from '../../../graphql/AllIdeasQuery.graphql';
import DebateThematicsQuery from '../../../graphql/DebateThematicsQuery.graphql';
import { PHASES } from '../../../constants';

const queries = {
  [PHASES.survey]: DebateThematicsQuery,
  default: AllIdeasQuery
};

export function prefetchMenuQuery(
  client: ApolloClient,
  variables: { identifier: string, discussionPhaseId: string, lang: string }
) {
  const query = queries[variables.identifier];
  client.query({
    query: query || queries.default,
    variables: variables
  });
}

type MenuTableProps = {
  identifier: string,
  phaseId: string
};

function MenuTable(props: MenuTableProps) {
  const { identifier, phaseId } = props;
  const discussionPhaseId = phaseId ? atob(phaseId).split(':')[1] : null;
  switch (identifier) {
  case PHASES.survey:
    return <SurveyTable {...props} discussionPhaseId={discussionPhaseId} />;
  case PHASES.voteSession:
    return null;
  default:
    return <IdeasTable {...props} discussionPhaseId={discussionPhaseId} />;
  }
}

export default MenuTable;