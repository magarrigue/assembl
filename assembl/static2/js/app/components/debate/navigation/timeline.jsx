// @flow
import * as React from 'react';
import { isCurrentPhase, getBarPercent, isStepCompleted } from '../../../utils/timeline';
import TimelineSegment from './timelineSegment';

type TimelineProps = {
  timeline: Timeline,
  showNavigation: boolean,
  activeSegment: number,
  onItemSelect: Function,
  onItemDeselect: Function
};

export function DumbTimeline({ timeline, activeSegment, showNavigation, onItemSelect, onItemDeselect }: TimelineProps) {
  return (
    <div className="timeline-container">
      {timeline &&
        timeline.map((phase, index) => (
          <TimelineSegment
            title={phase.title}
            index={index}
            active={index === activeSegment}
            key={index}
            barPercent={getBarPercent(timeline[index])}
            isCurrentPhase={isCurrentPhase(timeline[index])}
            showNavigation={showNavigation}
            phaseIdentifier={phase.identifier}
            phaseId={phase.id}
            startDate={phase.start}
            endDate={phase.end}
            isStepCompleted={isStepCompleted(phase)}
            onSelect={onItemSelect}
            onDeselect={onItemDeselect}
          />
        ))}
    </div>
  );
}

export default DumbTimeline;