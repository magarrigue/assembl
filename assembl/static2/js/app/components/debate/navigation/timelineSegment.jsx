// @flow
import * as React from 'react';
import { withApollo, type ApolloClient } from 'react-apollo';
import { Translate, Localize } from 'react-redux-i18n';
import { connect } from 'react-redux';
import classNames from 'classnames';

import { prefetchMenuQuery } from './menuTable';
import { getPhaseStatus, isSeveralIdentifiers } from '../../../utils/timeline';
import { displayModal } from '../../../utils/utilityManager';
import { get, goTo } from '../../../utils/routeMap';
import { isMobile } from '../../../utils/globalFunctions';
import { PHASE_STATUS, PHASES } from '../../../constants';

export const phasesToIgnore = [PHASES.voteSession];

export type DebateType = {
  debateData: {
    slug: string,
    useSocialMedia: boolean
  }
};

type TimelineSegmentProps = {
  timeline: Timeline,
  index: number,
  client: ApolloClient,
  title: string,
  startDate: string,
  endDate: string,
  phaseIdentifier: string,
  phaseId: string,
  debate: DebateType,
  barPercent: number,
  locale: string,
  active: boolean,
  onSelect: Function,
  onDeselect: Function
};

type TimelineSegmentState = {
  active: boolean
};

export class DumbTimelineSegment extends React.Component<TimelineSegmentProps, TimelineSegmentState> {
  state = {
    active: false
  };

  componentWillMount() {
    const { phaseId, phaseIdentifier, title, startDate, endDate, locale, client } = this.props;
    this.isTouchScreenDevice = isMobile.any();
    this.phaseStatus = getPhaseStatus(startDate, endDate);
    const notStarted = this.phaseStatus === PHASE_STATUS.notStarted;
    const ignore = phasesToIgnore.includes(phaseIdentifier);
    this.ignoreMenu = ignore && !notStarted;
    this.phaseName = title;
    const discussionPhaseId = phaseId ? atob(phaseId).split(':')[1] : null;
    if (discussionPhaseId && !ignore && !notStarted) {
      // don't prefetch if we're not going to use it
      prefetchMenuQuery(client, {
        lang: locale,
        identifier: phaseIdentifier,
        discussionPhaseId: discussionPhaseId
      });
    }
  }

  phaseStatus = null;

  phaseName = null;

  ignoreMenu = false;

  isTouchScreenDevice = false;

  segment = null;

  select = () => {
    const { onSelect, index } = this.props;
    onSelect(index);
  };

  renderNotStarted = () => {
    const { startDate } = this.props;
    return (
      <div>
        <Translate value="debate.notStarted" phaseName={this.phaseName} />
        <Localize value={startDate} dateFormat="date.format" />
      </div>
    );
  };

  displayPhase = () => {
    const { phaseId, phaseIdentifier, onDeselect, timeline } = this.props;
    const { debateData } = this.props.debate;
    const params = { slug: debateData.slug, phase: phaseIdentifier, phaseId: phaseId };
    const isSeveralPhases = isSeveralIdentifiers(timeline);
    if (isSeveralPhases) {
      if (this.phaseStatus === PHASE_STATUS.notStarted) {
        displayModal(null, this.renderNotStarted(), true, null, null, true);
        onDeselect();
      }
      if (this.phaseStatus === PHASE_STATUS.inProgress || this.phaseStatus === PHASE_STATUS.completed) {
        goTo(get('debate', params));
        onDeselect();
      }
    } else {
      goTo(get('debate', params));
      onDeselect();
    }
  };

  render() {
    const { barPercent, title, active } = this.props;
    const inProgress = this.phaseStatus === PHASE_STATUS.inProgress;
    const timelineClass = 'timeline-title txt-active-light';
    const touchActive = this.isTouchScreenDevice && !active;
    const onClick = touchActive ? this.select : this.displayPhase;
    return (
      <div
        ref={(segment) => {
          this.segment = segment;
        }}
        className={classNames('minimized-timeline', {
          active: active
        })}
        onMouseOver={!this.isTouchScreenDevice ? this.select : null}
      >
        <div onClick={onClick} className={timelineClass}>
          {inProgress && <span className="arrow assembl-icon assembl-icon-right-dir" />}
          <div className="timeline-link">{title}</div>
        </div>
        <div className="timeline-graph">
          <div className="timeline-bars">
            {barPercent > 0 && (
              <div className="timeline-bar-filler" style={barPercent < 20 ? { width: '20%' } : { width: `${barPercent}%` }}>
                &nbsp;
              </div>
            )}
            <div className="timeline-bar-background-container">
              &nbsp;
              <div className="timeline-bar-background" />
            </div>
          </div>
        </div>
        {!this.ignoreMenu && active && <span className="timeline-arrow" />}
      </div>
    );
  }
}

const mapStateToProps = state => ({
  locale: state.i18n.locale,
  debate: state.debate,
  timeline: state.timeline
});

// $FlowFixMe
export default connect(mapStateToProps)(withApollo(DumbTimelineSegment));