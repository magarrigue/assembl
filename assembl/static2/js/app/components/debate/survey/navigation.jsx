import React from 'react';
import { Grid, Col } from 'react-bootstrap';
import { connect } from 'react-redux';
import { Translate } from 'react-redux-i18n';
import { getDomElementOffset, calculatePercentage } from '../../../utils/globalFunctions';
import { getIfPhaseCompletedById } from '../../../utils/timeline';
import { SMALL_SCREEN_WIDTH } from '../../../constants';
import { withScreenDimensions } from '../../common/screenDimensions';

class Navigation extends React.Component {
  constructor(props) {
    super(props);
    const { questionsLength } = this.props;
    this.displayNav = this.displayNav.bind(this);
    this.displayPagination = this.displayPagination.bind(this);
    this.scrollToQuestion = this.scrollToQuestion.bind(this);
    this.state = {
      navPosition: 'fixed',
      navBottom: 0,
      isHidden: true,
      questionsLength: questionsLength,
      currentQuestionNumber: 0
    };
  }

  componentDidMount() {
    window.addEventListener('scroll', this.displayNav);
    window.addEventListener('scroll', this.displayPagination);
  }

  componentWillReceiveProps(nextProps) {
    this.setState(
      {
        questionsLength: nextProps.questionsLength
      },
      () => {
        this.displayNav();
        this.displayPagination();
        if (nextProps.isScroll && nextProps.questionIndex && this.props.screenWidth >= SMALL_SCREEN_WIDTH) {
          this.scrollToQuestion(nextProps.questionIndex);
        }
      }
    );
  }

  componentWillUnmount() {
    window.removeEventListener('scroll', this.displayNav);
    window.removeEventListener('scroll', this.displayPagination);
  }

  getQuestionsOffset(questionsLength) {
    const offsetArray = [];
    this.questionsLength = questionsLength;
    for (let i = 0; i < this.questionsLength; i += 1) {
      if (!document.getElementsByClassName('question-title')[i]) {
        return;
      }
      const questionOffset = Number(getDomElementOffset(document.getElementsByClassName('question-title')[i]).top);
      offsetArray.push(questionOffset + 40);
    }
    return offsetArray; // eslint-disable-line
  }

  displayNav() {
    const proposals = document.getElementById('proposals');
    if (!proposals) {
      return;
    }
    const proposalsOffset = getDomElementOffset(proposals).top;
    const firstTextarea = document.getElementsByClassName('rich-text-editor')[0];
    if (!firstTextarea) {
      return;
    }
    const limitToHide = getDomElementOffset(firstTextarea).top + firstTextarea.clientHeight;
    const limitToShow = limitToHide + document.getElementById('nav').clientHeight;
    const windowOffset = window.pageYOffset + this.props.screenHeight;
    const { timeline, phaseId } = this.props;
    const isPhaseCompleted = getIfPhaseCompletedById(timeline, phaseId);
    if (windowOffset < limitToHide && !isPhaseCompleted) {
      this.setState({
        isHidden: true
      });
    }
    if (windowOffset > limitToShow && !isPhaseCompleted) {
      this.setState({
        isHidden: false
      });
    }
    if (windowOffset > proposalsOffset) {
      this.setState({
        navPosition: 'absolute',
        navBottom: proposals.clientHeight,
        isHidden: true
      });
    } else {
      this.setState({
        navPosition: 'fixed',
        navBottom: 0
      });
    }
  }

  displayPagination() {
    const questionsOffset = this.getQuestionsOffset(this.state.questionsLength);
    if (!questionsOffset) {
      return;
    }
    const windowOffset = window.pageYOffset + this.props.screenHeight;
    let currentQuestionNumber = 0;
    for (let i = 0; i < this.state.questionsLength; i += 1) {
      if (windowOffset > questionsOffset[i]) {
        currentQuestionNumber = i + 1;
      }
    }
    this.setState({
      currentQuestionNumber: currentQuestionNumber
    });
  }

  scrollToQuestion(questionIndex) {
    let target;
    if (questionIndex > this.state.questionsLength) {
      target = document.getElementById('proposals');
    } else {
      target = document.getElementById(`q${questionIndex}`);
    }
    const targetOffset = Number(getDomElementOffset(target).top);
    window.scrollTo({ top: targetOffset - 80, left: 0, behavior: 'smooth' });
    this.props.scrollToQuestion(false);
  }

  render() {
    const barWidth = calculatePercentage(this.state.currentQuestionNumber, this.state.questionsLength);
    return (
      <section
        className={this.state.isHidden ? 'hidden' : 'shown navigation-section background-color'}
        id="nav"
        style={{ position: this.state.navPosition, bottom: this.state.navBottom }}
      >
        <Grid fluid>
          <div className="max-container">
            <div className="question-nav">
              <Col xs={12} md={9} className="col-centered">
                <Col xs={6} md={6} className="no-padding">
                  <div className="question-numbers">
                    <div className="txt">
                      <Translate
                        value="debate.survey.question_x_on_total"
                        current={this.state.currentQuestionNumber}
                        total={this.state.questionsLength}
                      />
                    </div>
                    <div className="bar" style={{ width: `${barWidth}%` }}>
                      &nbsp;
                    </div>
                    <div className="bkg-bar">&nbsp;</div>
                  </div>
                </Col>
                <Col xs={6} md={6} className="no-padding">
                  <div
                    className="arrow right"
                    onClick={() => {
                      this.scrollToQuestion(this.state.currentQuestionNumber + 1);
                    }}
                  >
                    <span className="assembl-icon-down-open" />
                  </div>
                  {this.state.currentQuestionNumber > 1 && (
                    <div
                      className="arrow right"
                      onClick={() => {
                        this.scrollToQuestion(this.state.currentQuestionNumber - 1);
                      }}
                    >
                      <span className="assembl-icon-up-open" />
                    </div>
                  )}
                </Col>
              </Col>
            </div>
          </div>
        </Grid>
      </section>
    );
  }
}

const mapStateToProps = state => ({
  timeline: state.timeline
});

export default withScreenDimensions(connect(mapStateToProps)(Navigation));