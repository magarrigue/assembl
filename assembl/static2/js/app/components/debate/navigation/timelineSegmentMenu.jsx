// @flow
import React from 'react';
import { Translate, Localize } from 'react-redux-i18n';

import MenuTable from './menuTable';
import { getPhaseStatus } from '../../../utils/timeline';
import { PHASE_STATUS } from '../../../constants';
import { phasesToIgnore } from './timelineSegment';

type TimelineSegmentMenuProps = {
  title: string,
  startDate: string,
  endDate: string,
  phaseIdentifier: string,
  phaseId: string,
  onMenuItemClick: Function
};

export function DumbTimelineSegmentMenu({
  phaseIdentifier,
  phaseId,
  onMenuItemClick,
  startDate,
  endDate,
  title
}: TimelineSegmentMenuProps) {
  const phaseStatus = getPhaseStatus(startDate, endDate);
  const isNotStarted = phaseStatus === PHASE_STATUS.notStarted;
  if (isNotStarted) {
    return (
      <div className="menu-container">
        <div className="not-started">
          <Translate value="debate.notStarted" phaseName={title} />
          <Localize value={startDate} dateFormat="date.format" />
        </div>
      </div>
    );
  }
  const ignoreMenu = phasesToIgnore.includes(phaseIdentifier);
  return !ignoreMenu ? (
    <div className="menu-container">
      <MenuTable identifier={phaseIdentifier} phaseId={phaseId} onMenuItemClick={onMenuItemClick} />
    </div>
  ) : null;
}

export default DumbTimelineSegmentMenu;