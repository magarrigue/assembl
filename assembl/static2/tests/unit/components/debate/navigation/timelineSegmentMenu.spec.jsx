import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';

import { DumbTimelineSegmentMenu } from '../../../../../js/app/components/debate/navigation/timelineSegmentMenu';

describe('DumbTimelineSegmentMenu component', () => {
  it('should match the DumbTimelineSegmentMenu', () => {
    const props = {
      title: {
        entries: [{ en: 'Foo' }]
      },
      startDate: 'date1',
      endDate: 'date2',
      phaseIdentifier: 'foo',
      phaseId: 'phaseFoo',
      locale: 'en'
    };
    const renderer = new ShallowRenderer();
    renderer.render(<DumbTimelineSegmentMenu {...props} />);
    const rendered = renderer.getRenderOutput();
    expect(rendered).toMatchSnapshot();
  });
});