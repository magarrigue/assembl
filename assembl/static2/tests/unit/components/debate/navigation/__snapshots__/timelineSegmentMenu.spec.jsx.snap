// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DumbTimelineSegmentMenu component should match the DumbTimelineSegmentMenu 1`] = `
<div
  className="menu-container"
>
  <MenuTable
    identifier="foo"
    onMenuItemClick={undefined}
    phaseId="phaseFoo"
  />
</div>
`;
